Fixes issue2251: Only sign block after SCP externalize

hewison-chris opened the pull request:
When the SCP balloting completes and the single event EXTERNALIZE is triggered then we can sign the block and gossip the signature to other nodes. This is to prevent leaking our private key by signing multiple blocks due to more than one COMMIT ballot.

First two commits have been put in seperate PR https://github.com/bosagora/agora/pull/2302

hewison-chris committed:
Only sign block after SCP externalize

When the SCP balloting completes and the single event EXTERNALIZE is
triggered then we can sign the block and gossip the signature to other
nodes. This is to prevent leaking our private key by signing multiple
blocks due to more than one COMMIT ballot.

Geod24 commented:
I assume this is still a WIP ?

hewison-chris commented:
> I assume this is still a WIP ?

Yes, I am still fixing tests which now fail due to the changes so far. Also we need to consider the signing during catchup potential problems.

hewison-chris commented:
Now :green_circle: and last commit prevents signing during block catch-up.

Geod24 commented on source/agora/test/PeriodicCatchup.d:
```diff
@@ -94,7 +94,7 @@ private class NodeManager (): TestAPIManager
         {
             assert(conf.validator.enabled);
-            log.trace("Create node {} as cathup test node", this.nodes.length);
+            log.trace("Create node #{} ({}) as cathup test node", this.nodes.length, conf.validator.key_pair.address);
```
Own commit

hewison-chris replied:
This was extracted into separate commit and merged in PR https://github.com/bosagora/agora/pull/2288

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -21,7 +21,6 @@
 module agora.node.Ledger;
 
 import agora.common.Amount;
-import agora.common.BitMask;
```
I believe the import was already dead before ? If so, own commit.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -857,17 +856,6 @@ public class Ledger
         }
 
         assert(validators.length == block.header.validators.count);
-        // Check that more than half have signed
-        auto signed = block.header.validators.setCount;
-        if (signed <= validators.length / 2)
-        {
-            log.error("Block#{}: Signatures are not majority: {}/{}, signers: {}",
-                      block.header.height, signed, validators.length,
-                      block.header.validators);
-            return "The majority of validators hasn't signed this block";
-        }
-
-        log.trace("Checking signature, participants: {}/{}", signed, validators.length);
```
I don't think that works, because now `validateBlock` will not check the number of signatures:
https://github.com/bosagora/agora/blob/11ce6e379fd7629e39d31ed80b9e1cc8e02a93cc/source/agora/node/Ledger.d#L746-L798

Which means that we will accept *any* blocks on our endpoints, doesn't it ?

hewison-chris replied:
I think we might need to be more strict about included signatures when validating blocks we receive. But when we are active in the nomination and externalise when informed by SCP then we can not be strict about % of signatures as we need to add them afterwards. I'll try and implement this.

hewison-chris replied:
I've added an array to the `ValidatingLedger` to keep track of blocks we are `EXTERNALIZE`ing from `SCP` event. These we can ignore the majority check but for all other cases we want to validate the majority and reject. I think this also covers the case where a `ValidatingLedger` is receiving blocks from the network during catchup after a restart.

Geod24 commented on source/agora/node/Validator.d:
```diff
@@ -282,66 +282,6 @@ public class Validator : FullNode, API
         return id;
     }
 
-    /***************************************************************************
-
-        Store the block in the ledger if valid.
-        If block is not yet signed by this node then sign it and also
-        gossip the block sig from this node to other nodes.
-
-        Params:
-            block = block to be added to the Ledger
-
-    ***************************************************************************/
```
https://github.com/bosagora/agora/blob/11ce6e379fd7629e39d31ed80b9e1cc8e02a93cc/source/agora/node/FullNode.d#L541-L549

Probably needs an update as well.

hewison-chris replied:
Good catch, I'll update.

hewison-chris replied:
Updated

Geod24 reviewed (changes requested):
I think the first commit is no bueno.

Geod24 commented:
Looks like there's still an issue ?

hewison-chris commented:
> Looks like there's still an issue ?

Yes will take a look.

hewison-chris commented:
> > Looks like there's still an issue ?
> 
> Yes will take a look.

The first problem is that the message which tells us why the `acceptBlock` failed is lost as it is assumed we can get the same failure by calling `validateBlock` (which is only failure case). However, the second call doesn't always fail or it could be a different reason so it is best to get the reason when it occurs. I added a commit for this change. Now I can debug the second issue where it is still checking the number of signatures when I think it should not in the `LocalTransaction` unit test.

hewison-chris committed:
`TestConf`: use `Consensus` default of 144 for `payout_period`

Since the catchup for missing signatures was limited to the last payment
block some tests started to fail occasionally. This commit will set the
default in testConf to a high enough value to prevent flaky tests.

hewison-chris commented:
The unit test `EmptyBlocks` failed as it was using the default `payment_period` of 5 which means it will sometimes fail. I added a commit to this PR to set default in test `Base.d` of 200 so that new tests will not by default be flaky.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1964,8 +1964,10 @@ public struct TestConf
 
-    /// How frequent the payments to Validators will be
-    public uint payout_period = 5;
+    /// How frequent the payments to Validators will be. If too low then tests
+    /// which assert all blocks are the same will sometimes fail if the catchup
+    /// is after height goes beyond a payment block.
+    public uint payout_period = 200;
```
The default should be the `ConsensusParams`'s default, that is, 144 (`ConsensusParams.init.payout_period`)

hewison-chris replied:
Updated.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -854,6 +811,11 @@ extern(D):
         in Hash block_hash) @safe nothrow
     {
         const PublicKey K = block_sig.public_key;
+        if (block_sig.height in this.slot_sigs && K in this.slot_sigs[block_sig.height])
+        {
+            log.trace("Signature already collected for public_key {} at height {}", block_sig.public_key, block_sig.height);
+            return false;
+        }
```
```D
if (auto sigs = block_sig.height in this.slot_sigs)
{
  if (K in *sigs)
  {
    /* ... */
  }
}
```

Regarding the message, perhaps use "for ourselves" ?

Geod24 replied:
Also this should really use the UTXO instead of the public key

Geod24 replied:
Actually:
```D
auto sigs = block_sig.height in this.slot_sigs:
if (sigs is null || K in *sigs) {}
```

And remove the redundant lookups below.

hewison-chris replied:
Updated to use:
```D
if (sigs is null || K in *sigs)
{
    log.trace("Signature already collected for this node at height {}", block_sig.height);
    return false;
}
```

hewison-chris replied:
I will raise an issue to tackle to use `UTXO` rather than `PublicKey`  in `slot_sigs`.

hewison-chris replied:
In fact we only need to check if this received signature is stored in the `slot_sigs` or not. We don't need the other signatures at this height. Will revert to use:-
```D
if (block_sig.height in this.slot_sigs && K in this.slot_sigs[block_sig.height])
```

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -977,30 +939,19 @@ extern(D):
+            // Now we add our signature and gossip to other nodes
+            log.trace("ADD BLOCK SIG at height {} for this node {}", height, this.kp.address);
+            this.slot_sigs[height][this.kp.address] = this.createBlockSignature(block);
+            gossipBlockSignature(ValidatorBlockSig(height, this.kp.address,
```
`this.`

hewison-chris replied:
Updated

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1370,6 +1392,31 @@ public class ValidatingLedger : Ledger
             clock, block_timestamp_tolerance, onAcceptedBlock);
     }
 
+    // dynamic array to keep track of blocks we are externalizing so can allow
+    //  less signatures than majority when validating
+    // TODO: We need to clear out old entries to reduce memory footprint
+    private Height[] externalizing;
+
+    public void addHeightAsExternalizing (Height height) @safe nothrow
+    {
+        externalizing ~= height;
+    }
+
+    /***************************************************************************
+
+        Used to indicate if the block we are validating has been triggered to
+            externalize by the SCP event EXTERNALIZE.
+
+        Params:
+            height = ignored here but used in ValidatingLedger function override
+
+    ***************************************************************************/
+
+    protected override bool isExternalizing (in Height height) @safe nothrow
+    {
+        return externalizing.canFind(height);
+    }
+
```
Well, first, the Ledger is not guaranteed to be a `ValidatingLedger`. So having a `isExternalizing` in the base class and overriding it here is still leaking abstraction. Overriding `validateBlockSignature` seems more sensible.

Also, even if we're going with this, is there a situation in which we would sign a block without signing intermediate blocks ? As in, do we ever sign block `N + 1` before block `N` ?

hewison-chris replied:
> Well, first, the Ledger is not guaranteed to be a `ValidatingLedger`. So having a `isExternalizing` in the base class and overriding it here is still leaking abstraction. Overriding `validateBlockSignature` seems more sensible.

I'll fix the leaking abstraction.  did consider overriding `validateBlockSignature` but I think it ended up with some duplication. Maybe a better fit is to have a `checkMajority` function which is overridden in the `ValidatingLedger`.

hewison-chris replied:
> Also, even if we're going with this, is there a situation in which we would sign a block without signing intermediate blocks ? As in, do we ever sign block `N + 1` before block `N` ?

The problem is not if we have signed the block, but if enough of the others have signed and we have received them. If we `externalize` blocks quickly we can be waiting on more than one block to have more than half signatures added.

Geod24 replied:
> If we externalize blocks quickly we can be waiting on more than one block to have more than half signatures added.

Do we externalize blocks on top of blocks for which we don't have a majority signature ?

hewison-chris replied:
We do if we are a `Validator` who is `externalizing` blocks due to `SCP` `EXTERNALIZE` events. Other nodes will not `externalize` blocks until they have the required majority of signatures. I think this is ok if we are relying on the fact that `SCP` `EXTERNALIZE` event can only happen once per block height. Also, I am not sure we can avoid this if we want regular block intervals.

Geod24 commented:
ENEEDREBASE again, sorry!

hewison-chris commented:
> ENEEDREBASE again, sorry!

Done

hewison-chris commented:
@Geod24 - can we merge this yet?

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -852,12 +852,8 @@ public class Ledger
-                      block.header.validators);
-            return "The majority of validators hasn't signed this block";
-        }
+            if (auto fail_msg = handleNotSignedByMajority(block.header, validators))
```
`this.`

hewison-chris replied:
Updated.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -845,8 +802,8 @@ extern(D):
-        Collect the block signature for a confirmed ballot or gossiped signature
-        only if the signature is valid for validator and block hash
+        Collect the block signature for a gossiped signature only if the
+            signature is valid for validator and block hash
```
Extra ws

hewison-chris replied:
Removed

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1370,6 +1386,40 @@ public class ValidatingLedger : Ledger
+
+    public void addHeightAsExternalizing (Height height) @safe nothrow
+    {
+        externalizing ~= height;
```
`this.`

hewison-chris replied:
Updated

Geod24 reviewed (approved):
Just a few styles nit.
